core/trino-main/src/main/java/io/trino/execution/scheduler/ScaledWriterScheduler.java

```diff
@@ -34,26 +35,29 @@
 import static io.trino.execution.scheduler.ScheduleResult.BlockedReason.WRITER_SCALING;
 import static io.trino.spi.StandardErrorCode.NO_NODES_AVAILABLE;
 import static io.trino.util.Failures.checkCondition;
+import static java.lang.Double.max;
 import static java.util.Objects.requireNonNull;
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
 
 public class ScaledWriterScheduler
         implements StageScheduler
 {
     private final StageExecution stage;
-    private final Supplier<Collection<TaskStatus>> sourceTasksProvider;
+    private final Supplier<Collection<TaskInfo>> sourceTasksProvider;
     private final Supplier<Collection<TaskStatus>> writerTasksProvider;
     private final NodeSelector nodeSelector;
     private final ScheduledExecutorService executor;
     private final long writerMinSizeBytes;
     private final int maxTaskWriterCount;
-    private final Set<InternalNode> scheduledNodes = new HashSet<>();
     private final AtomicBoolean done = new AtomicBoolean();
     private volatile SettableFuture<Void> future = SettableFuture.create();
 
+    @VisibleForTesting
+    final Set<InternalNode> scheduledNodes = new HashSet<>();
```

Comment:
Instead of exposing this, maybe you can initialize `InMemoryNodeManager` with a bunch of mock internal nodes and then directly call `schedule()` in tests?

Reply:
And you can also implement your mock version of `TestingStageExecution` where `scheduleTask` returns some mock remote task.

Reply:
I think this will be much cleaner and you'll be able to do end-to-end testing of this class using the `schedule()` method.

Reply:
Personally, I don't think it's clear or proper. Since, using Mockito is not preferred in Trino, unit tests become a lot messier with custom mocks and more error prone due to codes other than what really need to be tests (in this case, `getNewTaskCount` function. BTW, if this is Trino's convention, let me try to follow.